### PR TITLE
Overlay PR

### PR DESCRIPTION
**Pull request by WSSDude420**
- added separate loggers for each mod (available to mods through spdlogobject)
- added logger for outputting to console
- added LuaSQLite3 (available to mods through sqlite3 object)
- added json.lua from rxi (available to mods through json object)
- added Overlay (new master class put in place of former Console)
- made former Console widget of Overlay
- made VKBindings system to manage key bindings globally (Hotkeys, main system, directly responsible for added hotkeys.json)
- made Hotkeys widget for Overlay to have some GUI for binds setup (Hotkeys, controls VKBindings from GUI)
- stripped all input utilities from sol_ImGui
- fix DSR issues
- made Settings widget for Overlay (GUI for main config.json, config.json should not be packaged with releases from now on!)
-  onConsoleOpen/Close renamed to onOverlayOpen/Close
- added registerHotkey(id:string, description:string, handler:function()) for registering hotkey in menu from mod
- fix InputText/InputTextMutliline returning fixed-size buffer full of garbage after resulting string, breaking some components and logic in scripts of not accounted for

### CODE REVIEW

File: src/Image.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ void Image::Initialize()
             else
             {
                 for (auto c : pdb_info->Guid)
-                    spdlog::info("{:X}", (uint32_t)c);
+                    Logger::ErrorToMainFmt("{:X}", (uint32_t)c);
```
Comment (yamashi) [inline src/Image.cpp:57]: I won't merge this stuff, this is no how logging should be done, the function names are way too specialized

File: src/Paths.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        if (std::filesystem::exists(VKBindingsPath) && !std::filesystem::file_size(VKBindingsPath))
+            std::filesystem::remove(VKBindingsPath);
+
+        ModsPath = CETPath / "mods";
+        std::filesystem::create_directories(ModsPath);
+
+        Initialized = true;
+    }
+    
+    static inline std::filesystem::path ExePath{ };
```
Comment (yamashi) [inline src/Paths.h:37]: This is not okay, anyone can modify the content and fuck up for the rest of the program, const correctness must be respected

File: src/dllmain.cpp [hunk 3]
```diff
@@ -26,68 +26,72 @@ static HANDLE s_modInstanceMutex = nullptr;
 
 static void Initialize(HMODULE mod)
 {
+
+    s_modInstanceMutex = CreateMutex(NULL, TRUE, _T("Cyber Engine Tweaks Module Instance"));
```
Comment (yamashi) [inline src/dllmain.cpp:30]: Too early.

File: src/scripting/LuaVM.cpp [hunk 4]
```diff
@@ -23,14 +23,20 @@ void LuaVM::Shutdown()
 
 LuaVM& LuaVM::Get()
 {
+    assert(s_pLuaVM);
     return *s_pLuaVM;
 }
 
+std::vector<VKBindInfo>& LuaVM::GetBinds()
```
Comment (yamashi) [inline src/scripting/LuaVM.cpp:30]: const correctness

File: src/scripting/ScriptContext.cpp [hunk 5]
```diff
@@ -16,10 +17,58 @@ ScriptContext::ScriptContext(sol::state_view aStateView, const std::filesystem::
...
-            m_onConsoleClose = aCallback;
+        else if(acName == "onToolbarOpen")
+            m_onToolbarOpen = aCallback;
+        else if(acName == "onToolbarClose")
+            m_onToolbarClose = aCallback;
+        else
+            Logger::ErrorToModsFmt("Tried to register unknown handler '{}'!", acName);
+    };
+
+    m_env["registerVKBind"] = [this](const std::string& acID, const std::string& acDescription, sol::table aVKBindCode, sol::function aCallback)
```
Comment (yamashi) [inline src/scripting/ScriptContext.cpp:28]: This wasn't what we talked about, also the name is too specific, users are not supposed to know how we implement that

File: src/scripting/ScriptStore.cpp [hunk 6]
```diff
@@ -2,35 +2,55 @@
...
+        if (!std::filesystem::exists(fPath / "init.lua"))
+        {
+            Logger::ToConsoleFmt("Ignoring directory which misses init.lua! ('{}')", fPathStr);
+            Logger::WarningToModsFmt("Ignoring directory which misses init.lua! ('{}')", fPathStr);
             continue;
+        }
 
-        auto name = relative(file.path(), cScriptsPath).string();
-
+        auto name = std::filesystem::relative(fPath, Paths::ModsPath).string();
```
Comment (yamashi) [inline src/scripting/ScriptStore.cpp:24]: No need for full namespace

File: src/scripting/ScriptStore.cpp [hunk 7]
```diff
@@ -2,35 +2,55 @@
...
-    for (const auto& file : std::filesystem::directory_iterator(cScriptsPath))
+    for (const auto& file : std::filesystem::directory_iterator(Paths::ModsPath))
     {
         if (!file.is_directory())
             continue;
 
-        if (!exists(file.path() / "init.lua"))
+        auto fPath = file.path();
+        auto fPathStr = fPath.string();
+        if (!std::filesystem::exists(fPath / "init.lua"))
```
Comment (yamashi) [inline src/scripting/ScriptStore.cpp:17]: No need for full namespace

File: src/toolbar/HelperWidgets.h [hunk 8]
```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+inline const char* GetSpecialKeyName(UINT aVKCode)
```
Comment (yamashi) [inline src/toolbar/HelperWidgets.h:3]: Why is all of this inlined?

File: src/toolbar/Toolbar.cpp [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        apThis->isClipped = false;
+        return ClipCursor(nullptr);
+    }
+
+    return 1;
+}
+
+void Toolbar::Hook()
+{
+    uint8_t* pLocation = FindSignature({
```
Comment (yamashi) [inline src/toolbar/Toolbar.cpp:202]: Why is the mouse clip in the toolbar?

**Review by yamashi — CHANGES_REQUESTED** [review]

File: src/scripting/ScriptContext.cpp (hunk 5 above)
Comment (WSSDude420) [reply src/scripting/ScriptContext.cpp]: well maybe resolved dunno :P

File: src/toolbar/HelperWidgets.h (hunk 8 above)
Comment (WSSDude420) [reply src/toolbar/HelperWidgets.h]: moved to cpp, also moved GetSpecialKeyName into VKBindings, as i think it is better place for this than was here

File: src/toolbar/Toolbar.cpp (hunk 9 above)
Comment (WSSDude420) [reply src/toolbar/Toolbar.cpp]: well, hopefully better now... 

dont know if you want to completely separate that widget stuff to its own class from this and let overlay own just that class...
or if more-or-less main issue was that it was called toolbar when toolbar was kind of small part of it

**Comment by WSSDude420** [comment]
Should be fully compatible with #403

**Comment by stevenhoving** [comment]
I love your work, but @WSSDude420 if may I give you some advice. If you would split this merge request into multiple smaller ones. Then it is going to be a lot easier for you to get it merged. It is also is a lot easier to review, and for you to easier to resolve the review comments.

**Comment by WSSDude420** [comment]
Yeah, I know.... That is how they usually start, but it tends to grow after a while cause it is being tested and fixed and expanded on the go on our modding Discord always :D

Yeah, I know... Bad excuse... Well, is how it is currently. Definitely would like to change this, I even started that way back when I first started collaborating here. Discord ruined me in this regard, lazy to keep 10 PRs synced somehow to have some proper test builds for some QA and preparations for future for modders there. Many things are currently changing, quite rapidly thanks to continuous feedback.

Well, feels like some CEO apology reading this back now... :D 
Anyway, taken into account, I'll stop being lazy cause yeah, is messy :)

**Review by yamashi — APPROVED** [review]

**Pull request merged**